lib/opam_solve.ml

```diff
@@ -59,43 +63,109 @@ module Opam_monorepo_context (Base_context : BASE_CONTEXT) :
+    }
+
+  let is_valid_candidate ~allow_jbuilder ~require_dune ~name ~version opam_file
+      =
+    (* this function gets called way too often.. memoize? *)
     let pkg = OpamPackage.create name version in
     let depends = OpamFile.OPAM.depends opam_file in
     let depopts = OpamFile.OPAM.depopts opam_file in
     let uses_dune =
       Opam.depends_on_dune ~allow_jbuilder depends
       || Opam.depends_on_dune ~allow_jbuilder depopts
     in
-    let summary = Opam.Package_summary.from_opam ~pkg opam_file in
+    let summary = Opam.Package_summary.from_opam pkg opam_file in
     Opam.Package_summary.is_base_package summary
     || Opam.Package_summary.is_virtual summary
-    || uses_dune
+    || (not require_dune) || uses_dune
+
+  let rec remove_opam_provided_from_formula opam_provided filtered_formula :
+      OpamTypes.filtered_formula =
+    match filtered_formula with
+    | OpamFormula.Atom (name, _) as atom -> (
+        match OpamPackage.Name.Set.mem name opam_provided with
+        | true ->
+            Logs.debug (fun l ->
+                l "Removed %a from formula as it is opam-provided"
+                  Opam.Pp.package_name name);
+            OpamFormula.Empty
+        | false -> atom)
+    | OpamFormula.And (left, right) -> (
+        let left = remove_opam_provided_from_formula opam_provided left in
+        let right = remove_opam_provided_from_formula opam_provided right in
+        match (left, right) with
+        | Empty, right -> right
+        | left, Empty -> left
+        | left, right -> OpamFormula.And (left, right))
+    | OpamFormula.Or (left, right) -> (
+        let left = remove_opam_provided_from_formula opam_provided left in
+        let right = remove_opam_provided_from_formula opam_provided right in
+        match (left, right) with
+        | Empty, right -> right
+        | left, Empty -> left
+        | left, right -> OpamFormula.Or (left, right))
+    | otherwise -> otherwise
+
+  let remove_opam_provided_from_dependencies opam_provided opam_file =
+    let depends = OpamFile.OPAM.depends opam_file in
+    let depends = remove_opam_provided_from_formula opam_provided depends in
+    OpamFile.OPAM.with_depends depends opam_file
 
-  let filter_candidates ~allow_jbuilder ~name versions =
+  let filter_candidates ~allow_jbuilder ~require_dune ~name versions =
     List.map
       ~f:(fun (version, result) ->
         match result with
         | Error r -> (version, Error (Base_rejection r))
         | Ok opam_file ->
-            if is_valid_candidate ~allow_jbuilder ~name ~version opam_file then
-              (version, Ok opam_file)
+            if
+              is_valid_candidate ~allow_jbuilder ~require_dune ~name ~version
+                opam_file
+            then (version, Ok opam_file)
             else (version, Error Non_dune))
       versions
 
-  let candidates { base_context; fixed_packages; allow_jbuilder } name =
+  let remove_opam_provided ~opam_provided versions =
```

Comment:
Probably worth making this a no-op if `opam_provided` is empty rather than iterate over all opam files unconditionally.